core/src/main/java/tc/oc/pgm/projectile/ProjectileMatchModule.java

```diff
@@ -88,32 +101,96 @@ && isValidProjectileAction(event.getAction(), projectileDefinition.clickAction))
 
       if (this.isCooldownActive(player, projectileDefinition)) return;
 
-      boolean realProjectile = Projectile.class.isAssignableFrom(projectileDefinition.projectile);
+      boolean realProjectile = false;
+      if (projectileDefinition.projectile instanceof ProjectileDefinition.ProjectileEntity.RealEntity) {
+        realProjectile = Projectile.class.isAssignableFrom(
+            ((ProjectileDefinition.ProjectileEntity.RealEntity) projectileDefinition.projectile).entityType
+        );
+      }
       Vector velocity =
           player.getEyeLocation().getDirection().multiply(projectileDefinition.velocity);
-      Entity projectile;
+      Entity projectile = null;
+      BlockEntity blockEntity = null;
       try {
         assertTrue(launchingDefinition.get() == null, "nested projectile launch");
         launchingDefinition.set(projectileDefinition);
         if (realProjectile) {
           projectile = player.launchProjectile(
-              projectileDefinition.projectile.asSubclass(Projectile.class), velocity);
+            (((ProjectileDefinition.ProjectileEntity.RealEntity) projectileDefinition.projectile).entityType)
+              .asSubclass(Projectile.class),
+              velocity
+          );
           if (projectile instanceof Fireball fireball && projectileDefinition.precise) {
             NMSHacks.NMS_HACKS.setFireballDirection(fireball, velocity);
           }
         } else {
-          if (FallingBlock.class.isAssignableFrom(projectileDefinition.projectile)) {
+          if ((projectileDefinition.projectile instanceof ProjectileDefinition.ProjectileEntity.RealEntity) &&
+                  FallingBlock.class.isAssignableFrom((((ProjectileDefinition.ProjectileEntity.RealEntity) projectileDefinition.projectile).entityType))) {
             projectile =
                 projectileDefinition.blockMaterial.spawnFallingBlock(player.getEyeLocation());
           } else {
-            projectile =
-                player.getWorld().spawn(player.getEyeLocation(), projectileDefinition.projectile);
+            Location loc = player.getEyeLocation();
+            if ((projectileDefinition.projectile instanceof ProjectileDefinition.ProjectileEntity.AbstractEntity) &&
+                    ((ProjectileDefinition.ProjectileEntity.AbstractEntity) projectileDefinition.projectile).entityType == ProjectileDefinition.ProjectileEntity.AbstractEntityType.BLOCK) {
+              loc.setPitch(0);
+              loc.setYaw(0);
+              blockEntity = ENTITIES.spawnBlockEntity(loc, projectileDefinition.blockMaterial);
+              projectile = blockEntity.entity();
+            } else if (projectileDefinition.projectile instanceof ProjectileDefinition.ProjectileEntity.RealEntity) {
+              projectile =
+                      player.getWorld().spawn(loc, ((ProjectileDefinition.ProjectileEntity.RealEntity) projectileDefinition.projectile).entityType);
+            }
```

Comment:
```suggestion
            } else if (projectileDefinition.projectile instanceof ProjectileDefinition.ProjectileEntity.RealEntity re) {
              projectile = player.getWorld().spawn(loc, re.entityType());
            }
```
really should use java 17's features in a bunch of these